File: scripts/benchmark.js
#!/usr/bin/env node
// Benchmark the saved best bot against two baselines:
//   1. Random bot       — measures absolute strength
//   2. Default-weights minimax — measures improvement from training
//
// Usage:
//   node scripts/benchmark.js
//   node scripts/benchmark.js --games 200 --depth 6
import { runGames }                        from '../src/simulation/runner.js';
import { makeMinimaxBot, DEFAULT_WEIGHTS } from '../src/bots/minimax.js';
import { randomBot }                       from '../src/bots/random.js';
import { loadBest }                        from '../src/io/persistence.js';
import { fmtWeights }                      from '../src/viz/metrics.js';

// ─── Config ───────────────────────────────────────────────────────────────────

const DEFAULTS = { games: 200, depth: 6 };

const cfg = parseArgs(process.argv);

// ─── Load best bot ────────────────────────────────────────────────────────────

const seed = loadBest(DEFAULT_WEIGHTS);

if (seed.generation === 0) {
  console.error('No trained weights found. Run train.js first.');
  process.exit(1);
}

console.log('Connect4 AI — Benchmark');
console.log('─'.repeat(60));
console.log(`  Loaded: gen ${seed.generation}  fitness ${seed.fitness.toFixed(4)}`);
console.log(`  Weights: ${fmtWeights(seed.weights)}`);
console.log(`  Games per matchup: ${cfg.games}  |  Search depth: ${cfg.depth}`);
console.log('─'.repeat(60));
console.log('');

const bestBot    = makeMinimaxBot(cfg.depth, seed.weights);
const defaultBot = makeMinimaxBot(cfg.depth, DEFAULT_WEIGHTS);

// ─── vs Random ───────────────────────────────────────────────────────────────

process.stdout.write(`Running ${cfg.games} games vs random bot ...`);
const start1   = Date.now();
const vsRandom = runGames(bestBot, randomBot, cfg.games);
const t1       = ((Date.now() - start1) / 1000).toFixed(1);
process.stdout.write(`  done (${t1}s)\n`);
printResult('Best bot vs Random', vsRandom, cfg.games);

// ─── vs Default weights ───────────────────────────────────────────────────────

process.stdout.write(`Running ${cfg.games} games vs default-weights minimax ...`);
const start2    = Date.now();
const vsDefault = runGames(bestBot, defaultBot, cfg.games);
const t2        = ((Date.now() - start2) / 1000).toFixed(1);
process.stdout.write(`  done (${t2}s)\n`);
printResult(`Best bot vs Default (depth ${cfg.depth})`, vsDefault, cfg.games);

// ─── Summary ─────────────────────────────────────────────────────────────────

console.log('─'.repeat(60));
const overallWinRate = (vsRandom.wins + vsDefault.wins) / (cfg.games * 2);
console.log(`Overall win rate: ${(overallWinRate * 100).toFixed(1)}%  ` +
            `(${vsRandom.wins + vsDefault.wins}W / ${cfg.games * 2} games)`);
console.log('');

// ─── Helpers ─────────────────────────────────────────────────────────────────

function printResult(label, { wins, losses, draws }, total) {
  const pct = n => `${(n / total * 100).toFixed(1)}%`.padStart(6);
  const num = n => String(n).padStart(4);
  console.log('');
  console.log(`  ${label}`);
  console.log(`    Wins   ${num(wins)}   / ${total}  (${pct(wins)})`);
  console.log(`    Losses ${num(losses)} / ${total}  (${pct(losses)})`);
  console.log(`    Draws  ${num(draws)}  / ${total}  (${pct(draws)})`);
  console.log('');
}

function parseArgs(argv) {
  const opts = { ...DEFAULTS };
  for (let i = 2; i < argv.length; i++) {
    switch (argv[i]) {
      case '--games': opts.games = int(argv[++i], '--games'); break;
      case '--depth': opts.depth = int(argv[++i], '--depth'); break;
      default:
        console.error(`Unknown argument: ${argv[i]}`);
        process.exit(1);
    }
  }
  return opts;
}

function int(val, flag) {
  const n = parseInt(val, 10);
  if (isNaN(n) || n < 1) { console.error(`${flag} requires a positive integer`); process.exit(1); }
  return n;
}

File: scripts/benchmark.py
#!/usr/bin/env python3
# Benchmark the saved best bot against two baselines:
#   1. Random bot          — measures absolute strength
#   2. Default-weights minimax — measures improvement from training
#
# Usage:
#   python3 scripts/benchmark.py
#   python3 scripts/benchmark.py --games 200 --depth 6 --workers 5
#
# NOTE: the if __name__ == '__main__' guard is required when using
# multiprocessing — without it, worker processes would re-execute this
# script and spawn infinitely.
import sys
import time
import argparse
import multiprocessing
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.simulation.runner import run_games
from src.bots.minimax import make_minimax_bot, DEFAULT_WEIGHTS
from src.bots.random import random_bot
from src.io.persistence import load_best
from src.viz.metrics import fmt_weights


# ─── Helpers ─────────────────────────────────────────────────────────────────

def _print_result(label, result, total):
    wins   = result['wins']
    losses = result['losses']
    draws  = result['draws']
    def pct(n): return f'{n / total * 100:.1f}%'.rjust(6)
    print('')
    print(f'  {label}')
    print(f'    Wins   {wins:4d}   / {total}  ({pct(wins)})')
    print(f'    Losses {losses:4d} / {total}  ({pct(losses)})')
    print(f'    Draws  {draws:4d}  / {total}  ({pct(draws)})')
    print('')


def main():
    # ─── Config ───────────────────────────────────────────────────────────────

    parser = argparse.ArgumentParser(description='Benchmark trained bot vs baselines.')
    parser.add_argument('--games',   type=int, default=200,
                        help='Games per matchup (default: 200)')
    parser.add_argument('--depth',   type=int, default=6,
                        help='Search depth (default: 6)')
    parser.add_argument('--workers', type=int, default=5,
                        help='Parallel worker processes (default: 5)')
    args = parser.parse_args()

    cfg = {'games': args.games, 'depth': args.depth, 'workers': args.workers}

    # ─── Load best bot ────────────────────────────────────────────────────────

    seed = load_best(DEFAULT_WEIGHTS)

    if seed['generation'] == 0:
        print('No trained weights found. Run train.py first.', file=sys.stderr)
        sys.exit(1)

    print('Connect4 AI — Benchmark')
    print('─' * 60)
    print(f"  Loaded: gen {seed['generation']}  fitness {seed['fitness']:.4f}")
    print(f"  Weights: {fmt_weights(seed['weights'])}")
    print(f"  Games per matchup: {cfg['games']}  |  Search depth: {cfg['depth']}"
          f"  |  Workers: {cfg['workers']}")
    print('─' * 60)
    print('')

    best_bot    = make_minimax_bot(cfg['depth'], seed['weights'])
    default_bot = make_minimax_bot(cfg['depth'], DEFAULT_WEIGHTS)

    # ─── vs Random ───────────────────────────────────────────────────────────

    print(f"Running {cfg['games']} games vs random bot ...", end='', flush=True)
    start1    = time.perf_counter()
    vs_random = run_games(best_bot, random_bot, cfg['games'])
    t1        = time.perf_counter() - start1
    print(f'  done ({t1:.1f}s)')
    _print_result('Best bot vs Random', vs_random, cfg['games'])

    # ─── vs Default weights ───────────────────────────────────────────────────

    print(f"Running {cfg['games']} games vs default-weights minimax ...", end='', flush=True)
    start2     = time.perf_counter()
    vs_default = run_games(best_bot, default_bot, cfg['games'])
    t2         = time.perf_counter() - start2
    print(f'  done ({t2:.1f}s)')
    _print_result(f"Best bot vs Default (depth {cfg['depth']})", vs_default, cfg['games'])

    # ─── Summary ─────────────────────────────────────────────────────────────

    print('─' * 60)
    total_wins   = vs_random['wins'] + vs_default['wins']
    overall_rate = total_wins / (cfg['games'] * 2)
    print(f'Overall win rate: {overall_rate * 100:.1f}%  '
          f'({total_wins}W / {cfg["games"] * 2} games)')
    print('')


if __name__ == '__main__':
    multiprocessing.set_start_method('fork')
    main()
